### PR TITLE
SMHE-2011 Save firmware file to file system

### PR DESCRIPTION
**Pull request by jboon**
With this PR we removed the request back from dlms adapter to core in case a firmware file is not yet in the dlms adapter's cache repository when needed.
The firmware file content is no longer stored in the core database but stored by the core module on a configured file system. This file system is also reachable for the dlms adapter. 
Firmware file is still read once and stored in the cache repository.
ImageIdentifier is removed from the core database since this was a smartmetering specific solution which is no longer needed now the ImageIdentifier is stored on a file system in the same way as the firmware file.
A hash is stored in the core database to check the content of the firmware file stored in the file system. For this content check hash and hashtype are added to the firmware file properties in the firmware requests.

### CODE REVIEW

**Comment by harrymiddelburgetall** [comment]
✔ https://jenkins.fdp.osgp.cloud/view/SMHE-cucumber/job/gxf-smhe-cucumber-build/249/

File: integration-tests/cucumber-tests-platform-smartmetering/src/test/resources/features/osgp-adapter-ws-smartmetering/configuration/Firmware.feature [hunk 1]
```diff
@@ -219,24 +197,42 @@ Feature: SmartMetering Configuration - Firmware
...
+      | ManufacturerName | Kaifa |
+      | ModelCode        | MA10X |
     And a dlms device
       | DeviceIdentification | TEST1024000000002 |
       | DeviceType           | SMART_METER_E     |
       | ManufacturerCode     | KAI               |
       | DeviceModelCode      | MA105             |
-    And a firmware
-      | FirmwareFileIdentification  | TEST_FW_FILE_0001                        |
+#    And a firmware
```
Comment (stefanermens) [inline integration-tests/cucumber-tests-platform-smartmetering/src/test/resources/features/osgp-adapter-ws-smartmetering/configuration/Firmware.feature]: Can this block be removed?

File: osgp/platform/osgp-adapter-ws-core/src/main/java/org/opensmartgridplatform/adapter/ws/core/application/services/FirmwareManagementService.java [hunk 2]
```diff
@@ -676,10 +681,18 @@ private Manufacturer findManufacturerByCode(final String manufacturer)
...
   }
 
+  /**
+   * @param organisationIdentification
+   * @param firmwareFileAttributes
+   * @param file
+   * @param deviceModels
+   * @param firmwareModuleData
+   * @throws OsgpException
+   */
```
Comment (stefanermens) [inline osgp/platform/osgp-adapter-ws-core/src/main/java/org/opensmartgridplatform/adapter/ws/core/application/services/FirmwareManagementService.java]: This block doesn't add any information. Can it be removed?

File: osgp/platform/osgp-adapter-ws-core/src/main/java/org/opensmartgridplatform/adapter/ws/core/application/services/FirmwareManagementService.java [hunk 3]
```diff
@@ -689,14 +702,31 @@ public void addOrChangeFirmware(
...
+    this.firmwareFileRepository.save(firmwareFile);
+  }
+
+  /**
+   * Find for each DeviceModel from the WebServices the corresponding entity. There should be at
+   * least one DeviceModel. If none found a FunctionalException should be raised. Each DeviceModel
+   * can have its own Manufacturer (at least a theory) if a Manufacturer entity related to the
+   * DeviceModel can not be found a FunctionalException should be raised if one of the DeviceModel
+   * entities can not be found a FunctionalException should be raised.
+   */
```
Comment (stefanermens) [inline osgp/platform/osgp-adapter-ws-core/src/main/java/org/opensmartgridplatform/adapter/ws/core/application/services/FirmwareManagementService.java]: I guess some `.` are missing?

File: integration-tests/cucumber-tests-platform-common/src/test/java/org/opensmartgridplatform/cucumber/platform/common/glue/steps/ws/core/firmwaremanagement/FirmwareSteps.java [hunk 4]
```diff
@@ -180,8 +194,34 @@ protected void assertFirmwareFileExists(
...
+
+    final String fileStoreLocation =
+        getString(firmwareFileProperties, PlatformKeys.FIRMWARE_FILE_STORE_LOCATION, DATABASE);
+
+    if (fileStoreLocation == DATABASE) {
+      assertThat(databaseFirmwareFile.getFile()).isEqualTo(expectedFirmware.getFile());
+    }
+    if (fileStoreLocation == FILE_STORAGE) {
+      // There is no check here whether the file is actually on the file storage. The file storage
+      // check is implicitly done in the tests using the file from file storage.
```
Comment (stefanermens) [inline integration-tests/cucumber-tests-platform-common/src/test/java/org/opensmartgridplatform/cucumber/platform/common/glue/steps/ws/core/firmwaremanagement/FirmwareSteps.java]: I don't understand this comment. Is the file on the filestorage used or not?

File: integration-tests/cucumber-tests-platform-common/src/test/resources/features/common/osgp-adapter-ws-core/FirmwareManagement/AddOrChangeFirmware.feature [hunk 5]
```diff
@@ -8,25 +8,34 @@ Feature: FirmwareManagement add or change firmware
...
       | FirmwareDescription        | Add 1 - Firmware is newly created with File Contents |
       | FirmwareModuleVersionComm  | comm_1.2                                             |
+      | FirmwareFileStoreLocation  | <location>                                           |
+      | FirmwareHashType           | <hashtype>                                           |
+    Examples:
+      | imageidentifier  | file                                     | location     | hashtype | content | identifier |
+      | FEDCBA0987654321 | 57696520646974206c656573742069732067656b | FILE_STORAGE | SHA256   | with    | with       |
+      |                  | 57696520646974206c656573742069732067656b | FILE_STORAGE | SHA256   | with    | without    |
+      | FEDCBA0987654321 |                                          |              |          | without | with       |
+      |                  |                                          |              |          | without | without    |
```
Comment (stefanermens) [inline integration-tests/cucumber-tests-platform-common/src/test/resources/features/common/osgp-adapter-ws-core/FirmwareManagement/AddOrChangeFirmware.feature]: Location is always FILE_STORAGE (or null). There is no test with location DATABASE. Can the DATABASE option be removed from the FirmwareSteps?

File: integration-tests/cucumber-tests-platform-common/src/test/resources/features/common/osgp-adapter-ws-core/FirmwareManagement/AddOrChangeFirmware.feature [hunk 6]
```diff
@@ -8,25 +8,34 @@ Feature: FirmwareManagement add or change firmware
...
+      | FirmwareFileImageIdentifier | <imageidentifier>                                    |
     Then the add or change firmware response contains
       | Result | OK |
     And the firmware file '1234567890ABCBEF' exists
       | FirmwareFile               | 57696520646974206c656573742069732067656b             |
       | FirmwareFilename           | NewFirmware_test1                                    |
       | FirmwareDescription        | Add 1 - Firmware is newly created with File Contents |
       | FirmwareModuleVersionComm  | comm_1.2                                             |
+      | FirmwareFileStoreLocation  | <location>                                           |
+      | FirmwareHashType           | <hashtype>                                           |
```
Comment (stefanermens) [inline integration-tests/cucumber-tests-platform-common/src/test/resources/features/common/osgp-adapter-ws-core/FirmwareManagement/AddOrChangeFirmware.feature]: Only hashtype SHA256 is tested. Should there be a test for MD5 a well?

File: osgp/platform/osgp-adapter-ws-core/src/main/java/org/opensmartgridplatform/adapter/ws/core/application/services/FirmwareManagementService.java [hunk 7]
```diff
@@ -714,65 +744,75 @@ public void addOrChangeFirmware(
...
+      firmwareFile.setHash(fileDigest);
+      firmwareFile.setHashType(HashTypeDto.SHA256.name());
+      return firmwareFile;
+
+    } else {
+      return this.insertOrUdateDatabase(firmwareFileAttributes, file);
+    }
+  }
+
   private FirmwareFile insertOrUdateDatabase(
```
Comment (stefanermens) [inline osgp/platform/osgp-adapter-ws-core/src/main/java/org/opensmartgridplatform/adapter/ws/core/application/services/FirmwareManagementService.java]: `Update` instead of `Udate`

File: osgp/platform/osgp-adapter-ws-core/src/test/java/org/opensmartgridplatform/adapter/ws/core/application/services/FirmwareFileStorageServiceTest.java [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+  void storeFirmwareFileExisting() throws TechnicalException, IOException {
+    final String identification = "myFirmware";
+
+    this.service.storeFirmwareFile(FIRMWARE_FILE, identification);
+    this.service.storeFirmwareFile("some other file".getBytes(), identification);
+
+    final byte[] writtenFile = Files.readAllBytes(FIRMWARE_DIRECTORY.resolve(identification));
+    assertThat(writtenFile).isEqualTo(FIRMWARE_FILE);
+  }
+
```
Comment (stefanermens) [inline osgp/platform/osgp-adapter-ws-core/src/test/java/org/opensmartgridplatform/adapter/ws/core/application/services/FirmwareFileStorageServiceTest.java:88]: Do we need a test when the firmware file is null or empty?

File: osgp/platform/osgp-adapter-ws-core/src/test/java/org/opensmartgridplatform/adapter/ws/core/application/services/FirmwareManagementServiceTest.java [hunk 9]
```diff
@@ -0,0 +1,158 @@
...
+
+  @Mock private FirmwareModuleData firmwareModuleData;
+
+  @InjectMocks private FirmwareManagementService service;
+
+  @Captor ArgumentCaptor<FirmwareFile> fwfCaptor;
+
+  @BeforeEach
+  void setup() throws TechnicalException {
+    // Only testing the file storage solution
```
Comment (stefanermens) [inline osgp/platform/osgp-adapter-ws-core/src/test/java/org/opensmartgridplatform/adapter/ws/core/application/services/FirmwareManagementServiceTest.java]: Is there another solution?

File: osgp/platform/osgp-core/src/test/java/org/opensmartgridplatform/core/infra/jms/protocol/inbound/messageprocessors/GetFirmwareFileMessageProcessorTest.java [hunk 10]
```diff
@@ -66,7 +67,9 @@ public void processMessageShouldSendFirmwareFile() throws JMSException {
     final byte[] firmwareFileBytes = firmwareFileIdentification.getBytes();
 
     final UpdateFirmwareRequestDto updateFirmwareRequestDto =
-        new UpdateFirmwareRequestDto(firmwareFileIdentification, deviceIdentification);
+        new UpdateFirmwareRequestDto(
+            deviceIdentification,
+            new UpdateFirmwareRequestDataDto(firmwareFileIdentification, null, null));
```
Comment (stefanermens) [inline osgp/platform/osgp-core/src/test/java/org/opensmartgridplatform/core/infra/jms/protocol/inbound/messageprocessors/GetFirmwareFileMessageProcessorTest.java]: Should we test hashtype and digest as well?

**Review by stefanermens — CHANGES_REQUESTED** [review]

File: integration-tests/cucumber-tests-platform-smartmetering/src/test/resources/features/osgp-adapter-ws-smartmetering/configuration/Firmware.feature (hunk 1 above)
Comment (harrymiddelburgetall) [reply integration-tests/cucumber-tests-platform-smartmetering/src/test/resources/features/osgp-adapter-ws-smartmetering/configuration/Firmware.feature]: Yes, removed

File: integration-tests/cucumber-tests-platform-common/src/test/java/org/opensmartgridplatform/cucumber/platform/common/glue/steps/ws/core/firmwaremanagement/FirmwareSteps.java (hunk 4 above)
Comment (harrymiddelburgetall) [reply integration-tests/cucumber-tests-platform-common/src/test/java/org/opensmartgridplatform/cucumber/platform/common/glue/steps/ws/core/firmwaremanagement/FirmwareSteps.java]: Elaborated the comment a bit more

File: integration-tests/cucumber-tests-platform-common/src/test/resources/features/common/osgp-adapter-ws-core/FirmwareManagement/AddOrChangeFirmware.feature (hunk 5 above)
Comment (harrymiddelburgetall) [reply integration-tests/cucumber-tests-platform-common/src/test/resources/features/common/osgp-adapter-ws-core/FirmwareManagement/AddOrChangeFirmware.feature]: Commented in the feature file on why only FILE_STORAGE and SHA256 is tested.

File: integration-tests/cucumber-tests-platform-common/src/test/resources/features/common/osgp-adapter-ws-core/FirmwareManagement/AddOrChangeFirmware.feature (hunk 6 above)
Comment (harrymiddelburgetall) [reply integration-tests/cucumber-tests-platform-common/src/test/resources/features/common/osgp-adapter-ws-core/FirmwareManagement/AddOrChangeFirmware.feature]: Commented in the feature file on why only FILE_STORAGE and SHA256 is tested.

File: osgp/platform/osgp-adapter-ws-core/src/main/java/org/opensmartgridplatform/adapter/ws/core/application/services/FirmwareManagementService.java (hunk 7 above)
Comment (harrymiddelburgetall) [reply osgp/platform/osgp-adapter-ws-core/src/main/java/org/opensmartgridplatform/adapter/ws/core/application/services/FirmwareManagementService.java]: Corrected typo

File: osgp/platform/osgp-adapter-ws-core/src/main/java/org/opensmartgridplatform/adapter/ws/core/application/services/FirmwareManagementService.java (hunk 2 above)
Comment (harrymiddelburgetall) [reply osgp/platform/osgp-adapter-ws-core/src/main/java/org/opensmartgridplatform/adapter/ws/core/application/services/FirmwareManagementService.java]: Removed

File: osgp/platform/osgp-adapter-ws-core/src/main/java/org/opensmartgridplatform/adapter/ws/core/application/services/FirmwareManagementService.java (hunk 3 above)
Comment (harrymiddelburgetall) [reply osgp/platform/osgp-adapter-ws-core/src/main/java/org/opensmartgridplatform/adapter/ws/core/application/services/FirmwareManagementService.java]: Additional Sentence ending (.) and starting with capital clarifies the text

File: osgp/platform/osgp-adapter-ws-core/src/test/java/org/opensmartgridplatform/adapter/ws/core/application/services/FirmwareFileStorageServiceTest.java (hunk 8 above)
Comment (harrymiddelburgetall) [reply osgp/platform/osgp-adapter-ws-core/src/test/java/org/opensmartgridplatform/adapter/ws/core/application/services/FirmwareFileStorageServiceTest.java:88]: Added checks

File: osgp/platform/osgp-adapter-ws-core/src/test/java/org/opensmartgridplatform/adapter/ws/core/application/services/FirmwareManagementServiceTest.java (hunk 9 above)
Comment (harrymiddelburgetall) [reply osgp/platform/osgp-adapter-ws-core/src/test/java/org/opensmartgridplatform/adapter/ws/core/application/services/FirmwareManagementServiceTest.java]: SmartMetering application use the file storage solution. Other applications might use Database storage which is not covered by this test yet. Comment clarified.

File: osgp/platform/osgp-core/src/test/java/org/opensmartgridplatform/core/infra/jms/protocol/inbound/messageprocessors/GetFirmwareFileMessageProcessorTest.java (hunk 10 above)
Comment (harrymiddelburgetall) [reply osgp/platform/osgp-core/src/test/java/org/opensmartgridplatform/core/infra/jms/protocol/inbound/messageprocessors/GetFirmwareFileMessageProcessorTest.java]: The class GetFirmwareFileMessageProcessor tested here is not doing anything with these parameters. For completeness added to the test

**Review by stefanermens — APPROVED** [review]
